python/simulation.py

```diff
@@ -1897,6 +1894,41 @@ def load_chunk_layout(self, br, source):
             ## source is either filename (string)
             self.structure.load_chunk_layout(source, br)
 
+    def _get_load_dump_dirname(self, dirname, single_parallel_file):
+        """
+        Get the dirname to dump simulation state to.
+        """
+        if single_parallel_file:
+            dump_dirname = dirname
+        else:
+            # When doing a sharded dump (each process to its own file), use
+            # the process rank to get a unique name.
+            dump_dirname = os.path.join(dirname, 'rank%02d' % mp.my_rank())
+        return dump_dirname
+
+    def dump(self, dirname, structure=True, single_parallel_file=True):
+        """
+        Dumps simulation state.
+        """
+        dump_dirname = self._get_load_dump_dirname(dirname, single_parallel_file)
+        os.makedirs(dump_dirname, exist_ok=True)
```

Comment:
Note that the Meep module already contains a [`make_output_directory`](https://github.com/NanoComp/meep/blob/master/src/output_directory.cpp#L86-L130) function which is used by the unit tests, e.g. for a temp directory:

https://github.com/NanoComp/meep/blob/9509547b997747d9ba331c3a60a2508c8d918c67/python/tests/test_simulation.py#L25-L27

Reply:
When writing to a sharded file, each shard is written to its own sub-directory under the specified top-level directory - and hence the makedirs (see _get_load_dump_dirname)